addon/adapters/horizon.js

```diff
@@ -1,110 +1,160 @@
 import Ember from 'ember';
 import Adapter from 'ember-data/adapter';
 
+const { RSVP: {Promise}, get, inject: {service}, typeOf } = Ember;
+
 /**
  * @class HorizonAdapter
  *
- * This adapter uses Horizon's fetch (non-streaming) interface.
+ * This adapter integrates with a Horizon/RethinkDB stack; it provides
+ * normal CRUD operations by default and can provide real-time updates
+ * if the application chooses to use the services "watch" services for one,
+ * many, or all collections.
  */
 export default Adapter.extend({
-    _hzService: Ember.inject.service('horizon-connection'),
-
-    /**
-     * @method _getConnectionPromise
-     * @private
-     *
-     * Simple helper method to make sure horizon is connected and get the conn.
-     */
-    _getConnectionPromise() {
-        return this.get('_hzService').connect();
-    },
+    horizon: service(),
 
     // Horizon returns objects as plain JSON.
     defaultSerializer: 'json',
 
     findRecord(store, type, id) {
-        return this._getConnectionPromise().then((hz) => {
-            return new Ember.RSVP.Promise((resolve, reject) => {
-                hz(type.modelName)
-                    .find(id)
-                    .fetch()
-                    .subscribe(Ember.run.bind(this, resolve), Ember.run.bind(this, reject));
-            });
-        });
+      const horizon = get(this, 'horizon');
+      const subscription = horizon.getSubscription(type);
+      return new Promise((resolve, reject) => {
+
+        // If subscription exists we assume record already up-to-date
```

Comment:
1. This isn't the place for these checks, instead this should be incorporated into these existing adapter methods: https://github.com/emberjs/data/blob/0a2a7d5c6d568e145763292f872933291f19f950/addon/adapter.js#L463-L537
2. I would like to keep a differentiation between the fetch and watch adapters, this one is supposed to be the fetch adapter so I don't think that any kind of checking before fetching should be done at all in here. I'm going to go ahead and rename the file to avoid any confusion on which does which.